src/transfer.cpp

```diff
@@ -1199,4 +1252,431 @@ DirectReadSlot::~DirectReadSlot()
+        }
+        else
+        {
+            prevpriority = nextpriority - 2 * PRIORITY_STEP;
+        }
+    }
+    else
+    {
+        if (transfer->slot)
+        {
+            transfer_list::iterator cit = it + 1;
+            while (cit != transfers[transfer->type].end())
+            {
+                if (!(*cit)->slot)
+                {
+                    transfer->bt.arm();
+                    transfer->cachedtempurl = (*it)->slot->tempurl;
+                    delete transfer->slot;
+                    transfer->state = TRANSFERSTATE_QUEUED;
+                    break;
+                }
+
+                if (cit == dstit)
+                {
+                    break;
+                }
+
+                cit++;
+            }
+        }
+
+        prevpriority = (*dstit)->priority;
+        transfer_list::iterator nextit = dstit + 1;
+        nextpriority = (*nextit)->priority;
+    }
+
+    uint64_t newpriority = (prevpriority + nextpriority) / 2;
+    LOG_debug << "Moving transfer between priority " << prevpriority << " and " << nextpriority << ". New: " << newpriority;
+    if (prevpriority == newpriority)
+    {
+        LOG_warn << "There is no space for the move. Adjusting priorities.";
+        int positions = dstindex;
+        if (srcindex < dstindex)
+        {
+            positions++;
+        }
+
+        uint64_t fixedPriority = transfers[transfer->type][0]->priority - PRIORITY_STEP * (positions + 1);
+        for (int i = 0; i < positions; i++)
+        {
+            Transfer *t = transfers[transfer->type][i];
+            LOG_debug << "Adjusting priority of transfer " << i << " to " << fixedPriority;
+            t->priority = fixedPriority;
+            client->transfercacheadd(t);
+            client->app->transfer_update(t);
+            fixedPriority += PRIORITY_STEP;
+        }
+        newpriority = fixedPriority;
+        LOG_debug << "Fixed priority: " << fixedPriority;
+    }
+
+    transfer->priority = newpriority;
+    transfers[transfer->type].erase(it);
+    transfer_list::iterator fit = transfers[transfer->type].begin() + dstindex;
+    assert(fit == transfers[transfer->type].end() || (*fit)->priority != transfer->priority);
+    transfers[transfer->type].insert(fit, transfer);
+    client->transfercacheadd(transfer);
+    client->app->transfer_update(transfer);
+
+    if (up && !transfer->slot && !(client->moretransfers(transfer->type) && client->slotavail()))
+    {
+        transfer_list::iterator cit = iterator(transfer);
+        transfer_list::iterator it = transfers[transfer->type].end();
+        if (cit == it)
+        {
+            return;
+        }
+
+        it--;
+        while (it != cit)
+        {
+            if ((*it)->slot && (*it)->state == TRANSFERSTATE_ACTIVE)
+            {
+                (*it)->bt.arm();
+                (*it)->cachedtempurl = (*it)->slot->tempurl;
+                delete (*it)->slot;
+                (*it)->state = TRANSFERSTATE_QUEUED;
```

Comment:
Indeed, it's not strictly needed to save that state to the transfer cache in this moment. But it could cause bugs difficult to detect in the future, so I have added it.